lib/twitter-adapter.js

```diff
@@ -1,116 +1,182 @@
+        //
+        //    -from :0:    (-to :now:)  (pure historical)
+        //   (-from :now:)  -to :end:   (pure live)
+        //
+        if (from.unixms() === 0 && to.eq(params.now)) {
+            this.streaming = false;
+        } else if (from.eq(params.now) && to.isEnd()){
+            this.streaming = true;
+        } else {
+            throw this.compileError('INVALID-OPTION-COMBINATION', {
+                option: 'from/to',
+                rule: 'pure historical or pure live'
+            });
         }
-        this.query = params.filter_ast.expression.value;
 
-        this.do_streaming = options.stream || false;
-        this.total = options.total || 1000;
-        this.count = options.count || 100;
+        this.logger.debug('initializing in', this.streaming ? 'streaming' : 'historical', 'mode');
+
+        // Historical read limit
+        this.limit = options.limit || 1000;
 
-        this.delay = options.delay || JuttleMoment.duration(1, 's');
+        // Number to fetch for each historical read
+        this.fetchSize = options.fetchSize || 100;
+
+        // Live buffer size
+        this.bufferLimit = options.bufferLimit || 10000;
 
         // buffer of received tweets for search mode
         this.buffer = [];
 
         // live buffer for tweets in streaming mode
-        this.live = new Heap(function(p1, p2) {
+        this.reorderBuffer = new Heap(function(p1, p2) {
             return p1.time.lt(p2.time) ? -1 : 1;
         });
-    },
 
-    start: function() {
-        if (this.do_streaming) {
-            this.start_streaming();
-        } else {
-            this.search();
-        }
-    },
+        // points arriving before this time are dropped
+        this.reorderFrom = params.now.quantize(JuttleMoment.duration(1, 's'));
+
+        this.warnLate = this.throttledWarning('POINTS-ARRIVED-LATE', {
+            proc: 'read twitter'
+        });
+
+        this.warnOverflow = this.throttledWarning('BUFFER-LIMIT-EXCEEDED', {
+            buffer: 'read twitter buffer'
+        });
+        this.client = new Twitter(config);
+    }
+
+    static allowedOptions() {
+        return AdapterRead.commonOptions().concat(['limit', 'fetchSize', 'bufferLimit']);
+    }
+
+    defaultTimeOptions() {
+        return {
+            from: this.params.now,
+            to: this.params.now,
+            lag: JuttleMoment.duration(2, 's')
+        };
+    }
 
-    start_streaming: function() {
-        var self = this;
+    periodicLiveRead() {
+        return true;
+    }
 
-        this.live_window_start = new JuttleMoment().add(this.delay);
-        this.live_window_next = this.live_window_start.add(this.delay);
+    start() {
+        if (this.streaming) {
+            this._startStreaming();
+        }
+    }
 
-        this.program.scheduler.schedule(this.live_window_next.unixms(), function() { self.stream_points() });
+    read(from, to, limit, state) {
+        if (this.streaming) {
```

Comment:
Wouldn't it be safer to use `Promise.try` here instead of `promise.resolve` in `fetchLive` creating a new promise for `fetchHistorical`. If then at least the `new Promise` code could be in `_fetchHistorical` function?